Fix double stringification issue

cpachmayr opened the pull request:
## Summary

This PR fixes an issue where string values are being double-stringified in the `standardizeOutput` function, which causes problems when passing JSON strings between nodes.

When a string containing JSON is processed through the Code node (or any node using the `standardizeOutput` function), the string gets double-quoted, which can lead to invalid JSON errors when attempting to parse the string later in a workflow. This causes users to implement awkward workarounds, like manually removing extra quotes from stringified values.

The fix adds a type check to prevent strings from being JSON.stringify()'d while still properly handling other non-standard object types like Date and RegExp.

## Related Linear tickets, Github issues, and Community forum posts

Fixes #[13771](https://github.com/n8n-io/n8n/issues/13771)

## Review / Merge checklist

- [x] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md))
- [x] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created. <!-- No docs update needed for this bugfix -->
- [x] Tests included. <!-- The fix is covered by existing tests and we've added a specific test case for this scenario -->
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

cpachmayr committed:
fix(Code Node): Prevent double-stringification of string values

cpachmayr committed:
Add test case simulating real-world Code Node scenario to verify string output handling

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/n8n-io/n8n?pullRequest=13772) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/n8n-io/n8n?pullRequest=13772) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/n8n-io/n8n?pullRequest=13772) it.</sub>

Joffcom commented:
Hey @cpachmayr,

Thanks for the PR, We have created "GHC-1114" as the internal reference to get this reviewed.

One of us will be in touch if there are any changes needed, in most cases this is normally within a couple of weeks but it depends on the current workload of the team.

ShireenMissi commented on packages/nodes-base/nodes/Code/test/standardizeOutput.test.ts:
```diff
@@ -0,0 +1,185 @@
+		expect(result.data.specialChars).toContain('\n');
+	});
+
+	it('demonstrates the bug with pre-fix behavior (commented out)', () => {
```
Is there a reason this test is commented out? Should it be re-enabled or removed?

cpachmayr replied:
The commented-out test was left to illustrate the issue before the fix was added (more details in the comments (Lines 104-113) --- it's unnecessary for the fix; but also non-breaking to leave in.  I don't do many PRs, do you need me to remove and re-push with it removed, or are you able to just reject those lines on your end?

elsmr commented:
Hi @cpachmayr,

Thank you for your contribution!

Could you help me understand the problem you're trying to solve?

I have this testing workflow [here](https://gist.github.com/elsmr/74875783f24ee8e0a069bbec78b34309
) where I have a JSON string going into a Code node similar to your issue description.

I put a Set node after the Code node where I parse the JSON again and I don't run into any problem with double quotes.

elsmr commented:
@cpachmayr I tried with a Respond to Webhook node but the response still seems fine, no extra quotes and parses as JSON successfully.

Here's the webhook response:
<img width="1681" alt="image" src="https://github.com/user-attachments/assets/4f8d772a-3b8c-46a0-91dd-2f3874039d29" />
and the workflow [here](https://gist.github.com/elsmr/47bf2d0ce7a11c0cc34b1e5c00dcc4db)

Is it possible that this is only a visual problem where the output panel doesn't show escaped `"`? Seems the actual output data is fine.

If you can give me a reproduction workflow, I'm happy to have another look at it.

cpachmayr commented:
> @cpachmayr I tried with a Respond to Webhook node but the response still seems fine, no extra quotes and parses as JSON successfully.

> Is it possible that this is only a visual problem where the output panel doesn't show escaped `"`? Seems the actual output data is fine.
> 
> If you can give me a reproduction workflow, I'm happy to have another look at it.

I believe the detailed information [here](https://github.com/n8n-io/n8n/issues/13771) proves it isn't just a visual issue, at least on our system, setup, version, etc. The other problem is that my example workflow was a precisely just an attempt to show **_visually_** what was happening. 

The real problem, IMHO, is in `packages/nodes-base/nodes/Code/utils.ts` with the `standardOutput()` function and the recursion not checking if an item is already a string, which is where I think the focus should be. 

I don't do many PRs, and I can see now why it may have been better to give you the exact broken workflow rather than try only to provide the fix and visual proof. So my lesson learned, and I'll try to do that in the future. 😄 

Unfortunately, we changed the original workflow (we inserted a `Generate JSON` node rather than using our hack to strip the extra quotes). With only a one-day revision history, I can't provide the exact setup of the Respond to Webhook -- but here is more information: 

We stumbled onto the bug using  _Respond with: JSON Respond with a custom JSON body_. Before finding the bug, we sent plain text strings as the response. We discovered the bug when returning a custom JSON Body with three response variations with `plainText`, `markdown`, and `html` fields in the output. Pardon the pseudo-code for time:
```
{ "response": {
  "plainText": {{the plain text input}},
  "makdown": {{the markdown input}},
  "HTML": {{the HTML input}}
  }
}
```
The point is that when we moved past plain text, the workflow started breaking because the data needed to be stringified. So we tried stringifying the data. Initially, we tried doing JSON.stringify() on each field directly within the Custom JSON Body expression field. However, after reading the n8n documentation, I decided not to do that, and I pulled all of the stringification into a Code module before the Respond to Webhook node. 

The workflow then only broke intermittently, and it took us quite some time to find the source of the problem. The problem _only_ occurred when one of the inputs had items that needed to be escaped, and solving for that led us to the solution provided in the Bug report and our write-up for the PR.

If I can convince someone else to rebuild the original workflow, I'll pass it along—but I feel like [13771](https://github.com/n8n-io/n8n/issues/13771) shouldn't be closed without the PR. I'm happy to continue trying things or finding other ways to demonstrate the issue. 😄 I greatly appreciate your investigation into the code and matter.

cpachmayr commented:
Curious if anything else is needed on this. Finding similar issues with stringification in other areas and realize there's bigger issues to deal with -- but wanted to see if any progress on this before pursuing others. Thanks!

ShireenMissi commented:
> Curious if anything else is needed on this. Finding similar issues with stringification in other areas and realize there's bigger issues to deal with -- but wanted to see if any progress on this before pursuing others. Thanks!

Hi @cpachmayr Thanks for the PR and for reporting the issue! We’ve tried to reproduce the issue internally but couldn’t find any problems that need fixing. We believe the double quotes in your example are a result of using JSON.stringify(), and this behavior is expected.

The Code Node is used by a lot of our users, so we want to be cautious about introducing changes without a clear reason. If you have an example where the JSON is double quoted without explicitly using JSON.stringify(), please share it, and I’d be happy to take another look at the PR!
